contentcuration/contentcuration/frontend/shared/data/changes.js

```diff
@@ -124,220 +36,124 @@ export function withChangeTracker(callback, expiry = 10 * 1000) {
-   *
-   * @param {number} milliseconds
+   * Clean up the changes to avoid holding onto the data for too long
    */
-  renew(milliseconds) {
-    return getLocks({ tracker_id: this.id }).then(locks => {
-      if (!locks.length) {
-        throw new Error('Locks have expired');
-      }
-
-      const now = new Date();
-      return ignoreTransaction(() => {
-        return db[CHANGE_LOCKS_TABLE].bulkPut(
-          locks.map(lock => {
-            if (!lock.expiry) {
-              lock.expiry = now.getTime();
-            }
-            lock.expiry += milliseconds;
-            return lock;
-          })
-        ).then(() => this.setTimeout);
-      });
-    });
+  cleanUp() {
+    this._startingRev = null;
+    this._changes = null;
   }
 
   /**
    * Reverts all changes that have occurred while this was tracking
    *
    * @return {Promise}
    */
-  revert() {
-    if (this._isReverted) {
-      return Promise.resolve();
+  async revert() {
+    if (this.tracking || this.reverted) {
+      return;
     }
 
-    if (!this._isBlocking) {
-      throw new Error('Unable to revert changes without locks');
+    if (!this._changes || !this._changes.length) {
+      throw new Error('Unable to revert changes without tracking some changes');
     }
 
-    this.clearTimeout();
-
-    this._isReverted = true;
+    this.reverted = true;
 
     if (Dexie.currentTransaction) {
       if (Dexie.currentTransaction.source === REVERT_SOURCE) {
-        return Promise.resolve();
+        return;
       }
 
       // We're in the middle of a transaction, so just abort that
       Dexie.currentTransaction.abort();
-      return Promise.resolve();
+      return;
     }
 
-    return getLocks({ tracker_id: this.id })
-      .then(locks => {
-        if (!locks.length) {
-          throw new Error('Nothing to revert');
-        }
-
-        return promiseChunk(locks, 1, ([lock]) => {
-          return this.doRevert(lock);
-        });
-      })
-      .then(() => this.dismiss())
-      .catch(e => {
-        this.dismiss();
-        return Promise.reject(e);
-      });
+    await this.doRevert();
   }
 
-  doRevert(lock) {
-    // As of writing this, this table should be `__changesForSyncing` or `__treeChangesForSyncing`
-    const changeTable = db[lock.table_name];
-
-    // First, we use our changes table to find all the changes since the "lock" started
-    // using the `rev_start` which was set when we started tracking. The `rev` on the
-    // table auto increments
-    return db
-      .transaction('rw', lock.table_name, () => {
+  doRevert() {
+    // We'll go through each change one by one and revert each.
+    //
+    // R. Tibbles: I think this could be done in two queries (TODO)
+    return promiseChunk(this._changes.reverse(), 1, ([change]) => {
+      const table = db[change.table];
+      return db.transaction('rw', change.table, () => {
+        // This source inherits from `IGNORED_SOURCE` so this will be ignored
```

Comment:
Thinking about the future, if we are immediately sending changes to the backend using websockets, maybe we should be emitting 'reversions' as full changes that also get propagated to the backend, rather than trying to catch them before they're applied?

More like `git revert`ing a commit, rather than dropping a commit. The main challenge comes for things like publishing and syncing where there is not a clear path for what a 'restore' looks like.

Reply:
This is a great catch because by removing the change locks, I did change the behavior to be more like a `git revert`, but I did not update this logic accordingly! This logic should no longer use an ignored source for the updates. At the moment, we're only wrapping copy and move operations, so events like publishing and syncing shouldn't even occur here, but it would be good to be defensive against those things.